BibBuddy/app/src/main/java/de/bibbuddy/LibraryFragment.java

```diff
@@ -111,6 +116,219 @@ public boolean onOptionsItemSelected(MenuItem item) {
+      bibExportLibrary();
+    }
+  }
+
+  private void requestStoragePermission() {
+    if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
+        Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+      showRequestPermissionDialog();
+    } else {
+      requestPermissions(new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+          STORAGE_PERMISSION_CODE);
+    }
+  }
+
+  private void showRequestPermissionDialog() {
+    AlertDialog.Builder reqAlertDialog = new AlertDialog.Builder(getContext());
+    reqAlertDialog.setTitle(R.string.storage_permission_needed);
+    reqAlertDialog.setMessage(R.string.storage_permission_alert_msg);
+    reqAlertDialog.setPositiveButton(R.string.ok,
+        (dialog, which) -> ActivityCompat.requestPermissions(getActivity(),
+            new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+            STORAGE_PERMISSION_CODE));
+    reqAlertDialog.setNegativeButton(R.string.cancel,
+        (dialog, which) -> dialog.dismiss());
+    reqAlertDialog.create().show();
+  }
+
+  /**
+   * Callback method, that checks the result from requesting permissions.
+   *
+   * @param requestCode unique integer value for the requested permission
+   *                    This value is given by the programmer.
+   * @param permissions array of requested name(s)
+   *                    of the permission(s)
+   * @param grantResults grant results for the corresponding permissions
+   *                     which is either PackageManager.PERMISSION_GRANTED
+   *                     or PackageManager.PERMISSION_DENIED.
+   */
+  @Override
+  public void onRequestPermissionsResult(int requestCode,
+                                         String[] permissions, int[] grantResults) {
+
+    switch (requestCode) {
+
+      case STORAGE_PERMISSION_CODE:
+        if (grantResults.length > 0
+            && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+          bibExportLibrary();
+
+        } else {
+          Toast.makeText(getContext(), R.string.storage_permission_denied,
+              Toast.LENGTH_SHORT).show();
+        }
+        break;
+
+      default:
+    }
+  }
+
+  private void bibExportLibrary() {
+    String rootPathStr = Environment.getExternalStorageDirectory() + File.separator
+        + folderName + File.separator + fileName + fileTypeBib;
+    try {
+      File file = new File(rootPathStr);
+
+      if (file.exists()) {
+        file.delete();
+      }
+
+      file.createNewFile();
+      FileOutputStream out = new FileOutputStream(file);
+      out.flush();
+      out.close();
+
+      writeBibFile();
+
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+
+    Toast.makeText(getContext(),
+        getString(R.string.exported_file_stored_in) + '\n'
+            + File.separator + folderName + File.separator + fileName
+            + fileTypeBib, Toast.LENGTH_LONG).show();
+  }
+
+  private String retrieveBibContent() {
```

Comment:
Duplicate code see BookFragment

Reply:
This is because these two methods have to collect the data that should be stored in the exported .bib file. So, they aim to make the same things. The difference is that they take different information. I do not think the name will be a problem, since it is a private method and is only accessible in the current class. 

But I can change them then to retrieveBibContentLibrary and retrieveBibContentShelf.

Reply:
You don't need to change the names of the functions. What I meant was e.g. that the code from line 254 to 313 in LibraryFragment and the code from line 314 to 372 in BookFragment is the same. Is it possible for you to create a new class with functions for this to avoid duplicate code?